runtime/queries/downloads/downloads.go

```diff
@@ -0,0 +1,110 @@
+package downloads
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/csv"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strconv"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/queries"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	Limit            string = "limit"
+	Compression      string = "compression"
+	Format           string = "format"
+	Request          string = "request"
+	downloadPriority int    = 0
+)
+
+type DownloadHandler struct {
+	Runtime *runtime.Runtime
+}
+
+func (h *DownloadHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set("Content-Type", "text/csv")
+
+	limitString := req.URL.Query().Get(Limit)
+	var limit int
+	var err error
+	if limitString != "" {
+		limit, err = strconv.Atoi(limitString)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to parse request: %s", err), http.StatusBadRequest)
+			return
+		}
+	}
+
+	marshalled, err := base64.StdEncoding.DecodeString(req.URL.Query().Get(Request))
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse request: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	request := &runtimev1.DownloadLinkRequest{}
+	err = proto.Unmarshal(marshalled, request)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse request: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	switch v := request.Request.(type) {
+	case *runtimev1.DownloadLinkRequest_MetricsViewToplistRequest:
+		v.MetricsViewToplistRequest.Limit = int64(limit)
+		err = h.executeToplist(req.Context(), w, v.MetricsViewToplistRequest)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	default:
+		http.Error(w, fmt.Sprintf("Unsupported request type: %s", reflect.TypeOf(v).Name()), http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (h *DownloadHandler) executeToplist(ctx context.Context, writer http.ResponseWriter, req *runtimev1.MetricsViewToplistRequest) error {
```

Comment:
Since the export logic will be custom to each query, how about we implement it on the `Query` type instead? E.g. we could add a new func to the `Query` interface:

```go
	Export(ctx context.Context, rt *Runtime, instanceID string, priority int, format runtimev1.DownloadFormat, w io.Writer) error
```

Internally, it could decide whether to call `.Resolve` on itself, or whether to run an entirely different implementation (e.g. using DuckDB to build the CSV using the `COPY` statement).

Most queries will not support exports. We should just have a const error `runtime.ErrExportNotSupported` that they can return.

Reply:
Makes sense, though the long parameter list says we don't structure properly somewhere.

Reply:
Should we move `rt *runtime.Runtime, instanceID string, priority int` into `Context`?

Reply:
I agree the `Query` (and now `Export`) funcs are doing too much now. I don't think hiding the params in the context is the right way to clean it up, though. I have a few ideas for how we could simplify it, but let's keep it like this for now, and consider larger refactors afterwards.